### ISSUE TITLE
StoreKit testing: incorrect expiresDate. Timezone issue?

### ISSUE THREAD

**Issue by mobmad** [RevenueCat/purchases-ios#491]
**Environment**
   1. Platform: iOS
   2. SDK version: 3.10.7
   3. OS version: macOS 11.2.3 (20D91) / iOS 14.4 simulator
   4. Xcode version: 12.4 (12D4e)
   5. Affects at least StoreKit Testing. I have not tested against production environments.

**Steps to reproduce**
1. Follow the instructions in https://docs.revenuecat.com/docs/apple-app-store#ios-14-only-testing-on-the-simulator
2. In XCode, select `Configuration.storekit` -> Editor -> Time rate -> 1 second is one day
3. Build and run. Make a purchase. I used a Monthly subscription with a duration of 1 month, i.e. 30 days * 1 second = expiry in 30 seconds.
4. In XCode -> Manage StoreKit Transactions -> latest transaction -> "Subscription options" -> Cancel subscription

**Expected behaviour:**
After waiting 30 seconds (for the subscription to expire) + 5 minutes (or calling `invalidatePurchaserInfoCache()`, I tried both), `getPurchaserInfo` should reflect that the user doesn't have an active subscription anymore.

**Actual behaviour and bug:**
`getPurchaserInfo` returns an active subscription for **two hours** longer than it should, making me have to wait around two hours before I can test cancellation in the fastest scenario. This seems to be a timezone/GMT offset bug (see why I think so below).

**Things I've tried & found:**
* Retried calls to `getPurchaserInfo` at different intervals and waiting a lot longer that the 5 minute cache expiry.
* After confirming that `getPurchaserInfo` invalidates the cache (via debug logs) and refreshes from server, the **expired entitlement is still incorrectly returned**, i.e.:

```
activeEntitlements: {
    premium =     {
        billingIssueDetectedAt = null;
        expiresDate = "2021-04-08 11:44:13 +0000";
        isActive = Yes;
        isSandbox = Yes;
        latestPurchaseDate = "2021-04-08 11:43:43 +0000";
        originalPurchaseDate = "2021-04-08 09:29:59 +0000";
        periodType = null;
        productIdentifier = monthly;
        store = 0;
        unsubscribeDetectedAt = null;
        willRenew = Yes;
    };
}
```

**Note that my timezone is GMT+0200 (Central European Summer Time) and that the above output from `getPurchaserInfo` was made at 12:06 local time**. The entitlement should be expired, however it is not, and still returned under activeEntitlements. When looking up the same event via the RevenueCat dashboard I found something interesting (under Event Data):

```
{
  "app_user_id": "...",
  "currency": "USD",
  "entitlement_ids": [
    "premium"
  ],
  "environment": "SANDBOX",
  "event_timestamp_ms": 1617875024944,
  "expiration_at_ms": 1617882253000,
  "is_family_share": false,
  "is_trial_conversion": false,
  "price": 9.99,
  "price_in_purchased_currency": 9.99,
  "product_id": "monthly",
  "purchased_at_ms": 1617882223000,
  "store": "APP_STORE",
  "takehome_percentage": 0.85,
  "transaction_id": "StoreKitTest_Transaction_1617882223000_8"
}
```

Converting these timestamps to my local time yields:

* `event_timestamp_ms` = `Thu Apr 08 2021 11:43:44 GMT+0200 (Central European Summer Time)` (correct)
* `expiration_at_ms` = `Thu Apr 08 2021 13:44:13 GMT+0200 (Central European Summer Time)` (**incorrect, two hours off)**
* `purchased_at_ms` = `Thu Apr 08 2021 13:43:43 GMT+0200 (Central European Summer Time)` **(incorrect, two hours off)**

**To summarise:**
* RevenueCat dashboard/event data confirms my experience, making me have to wait two hours longer than expected for a subscription to expire
* `getPurchaserInfo` returns incorrect/inconsistent dates for (at least) `expiresDate` vs. the dashboard/API. Printing or inspecting the output from `getPurchaserInfo ` in XCode for `expiresDate` seems correct (see previous output higher up), but somewhere else in the SDK it seems to use the +2 hours offset data, since `activeEntitlements` is still mapped and returned, although being expired.

I can't tell if the bug is in the purchases-ios SDK, RevencueCat backend, or with StoreKit Test receipts, but I'm pretty sure there's a timezone issue causing this bug in one of those places.

**Comment by aboedo** [comment]
hi @mobmad! thanks for reporting, and for the detailed and thorough report. 

This issue is specific to StoreKit Test receipts. Thing is, they don't work like regular receipts - they can't be validated against `/verifyReceipts`, and instead have to be parsed on the backend... But the backend won't know about the changes you made in Xcode. 
So this is expected due to limitations in how these receipts work. It won't happen with regular Sandbox receipts or Production Receipts. 

Maybe if you try uninstalling the app, changing the duration from Xcode first, then installing and making a purchase, that'll generate a receipt with the new duration, but I'm not sure. 

I'm closing this issue since it's expected (albeit unfortunate) behavior, due to StoreKit Test receipts limitations. I'd encourage giving it a shot doing the updates from Xcode first and then a clean install, but I'm not certain it'll do the trick.

**Comment by mobmad** [comment]
Hi @aboedo! Thanks for looking into this quickly.

Uninstalling, installing and making a purchase yielded the same result, unfortunately.

So what you're saying is that this is an issue on Apples side for generating wrong timestamps in StoreKit Test receipts? Or that this is some limitation in RevenueCat's backend?

Either way; when parsing the receipt you know that it is a StoreKit Test receipt, right? So in theory, even if this is a bug on Apples side, you could detect that and fix that in either RevenueCat backend or iOS SDK?

I wouldn't close this issue as expected behaviour so quick, since:

1) StoreKit Testing seems to be the new way of testing going forward
2) Testing cancellations are important, and shouldn't block me for two hours between every test
3) Dealing with these low-level problems, so that we customers don't have to, is kind of RevenueCat's value proposition, right? IMO, this is certainly one of those problems I shouldn't have to deal with when using RevenueCat 🤞

**Comment by aboedo** [comment]
I apologize, @mobmad, I had misread the exact times you shared, and wrongfully assumed that you were strictly referring to the part where you can change the durations from Xcode, and figured that those changes might not reflect on the receipt until you reinstall. However, they should at most have a typical sandbox receipt duration, but never 2 hours. 

I think you may be correct about our backend assuming the wrong time zone for storekit test receipts. We'll do some digging to confirm which timezone the generated receipts use, and fix this if needed. 

Thanks for reporting and the persistence, and sorry for the confusion! I'm reopening this now, and I'll keep it open until a fix is deployed.

FWIW I fully agree with your points, again, I thought you were describing a different kind of issue.

**Comment by aboedo** [comment]
just confirmed this locally by updating the timezone on my device. We'll get this fixed. Thanks for reporting!!

**Comment by mobmad** [comment]
Ah, I see! Glad you're able to reproduce it and I'm looking forward to a fix :) Thanks again!

**Comment by aboedo** [comment]
Update here: we found the bug in the receipts generated by StoreKit Test. 
We're issuing a backend-side fix very soon, and we've also opened a radar so Apple can fix it on their side. Radar details: https://openradar.appspot.com/radar?id=5041399279910912

Thanks again for reporting! I'll close this issue when the fix is deployed.

**Comment by antoniobg** [comment]
Hi @mobmad, the backend fix is live now 🎊
I'm closing this issue, but let us know if you find any additional problem.

**Comment by mobmad** [comment]
Great! I can confirm that this works as expected now. 
Thanks again!
